### DISCUSSION TITLE
Make role names unique strings to enable similar features as "schema" for "permissions" collection

### DISCUSSION THREAD

**Discussion by cedenoVF** [directus/directus#20473] [Feature Requests]
### Current situation (V 10.7)

Currently, the roles in Directus are stored only as UUIDs, making it challenging to create backups or copy roles from one system to another (e.g., from a staging environment to a production environment). This limitation arises due to the use of UUIDs in the "permission" collection, which stores the details of each role's permissions.

GET /roles (response)
![image](https://github.com/directus/directus/assets/85623224/ce73d6e5-f304-4672-adfa-bf2cbff35859)

GET /permissions (response)
![image](https://github.com/directus/directus/assets/85623224/e1bd59a3-5e81-4c84-b28b-8cdf503d3564)

### Proposal

As a user (i.e. admin or IT maintainers of a CMS), we propose enhancing the role creation process in Directus by introducing unique names for roles. By allowing roles to have unique names, it would be possible to create backups and easily copy roles from one system to another without conflicts.

### Basic Example

Here's an example scenario to illustrate the benefits of this enhancement:

- In a staging environment, a new role called "Editor" is created and assigned specific permissions.
- With unique names for roles, the role "Editor" can be easily identified and copied to the production environment without conflicts.
- This enables seamless role synchronization across environments, reducing manual effort and minimizing the risk of errors.

Create an endpoint to compare two different versions of the permissions over the time, and create backups to restores permissions if something broke.

All features implemented for schema, supported in the permission collection https://docs.directus.io/reference/system/schema.html#schema

### Motivation

The rationale behind this proposed solution is to improve the overall flexibility and portability of role management in Directus. With unique names for roles, administrators would be able to easily identify and manage roles across different systems. This enhancement would streamline the deployment process and facilitate efficient role synchronization between environments.

In other words, if we have roles with name instead of UUID, then we can compare the roles for example of a **"production"** system with the **"stage"** system.

We could create templates, and test the directus _stage system_ and later copy those changes into the directus _production system_

The main reason is so we can do test on the API response in the stage with every change on the roles ( if any, for examples if new collections are created or new fields in a given collection). Check that the users in a given roles are getting exactly the data we have assigned via the UI. Then later take those changes and put it in the production system.

Currently, we will have to copy all roles of the stage system by hand to the production system or to create a workaround with some UUID matching based on the roles names.

Also we don't see any good reason to allow two roles with the same name:

![image](https://github.com/directus/directus/assets/85623224/55fe7d94-6af9-49c7-87a1-2c396166c284)

### Detailed Design

1. Enforce unique for names, so later any automatic matching of roles names with UUID from a test system to production, won't generate errors.

![image](https://github.com/directus/directus/assets/85623224/ce73d6e5-f304-4672-adfa-bf2cbff35859)

![image](https://github.com/directus/directus/assets/85623224/af5ea420-e19e-4ba1-b137-e486a5b49460)

2. Allows this to be setup at least with ENV variables if you don't want to disrupt any use case that is already doing so. i.e.
ENFORCE_ROLES_UNIQUE_NAME = TRUE

4. Provide for permissions similar ENDPOINTs as "https://docs.directus.io/reference/system/schema.html#schema"

- GET /permissions/snapshot
- POST /permissions/diff

etc.

### Must Have:

- Enable the creation of roles with unique names
- Ensure compatibility with existing systems and integrations
- Update the "permission" collection to accommodate the use of unique names

### Should Have:

- Provide a mechanism to easily map existing UUID-based roles to the new unique name-based roles
- Implement proper validation to prevent the creation of duplicate role names

### Could Have:

- Provide a migration script or tool to automatically update existing systems to support the new role creation process
- Allow administrators to customize the display names of roles while keeping the unique names intact

### Won't Have:

- Support for non-unique role names
- Major changes to the existing Directus architecture or functionalities

### Drawbacks

1. Users currently using the UUID and/or with duplicated roles name could be impacted
2. Users wont be able to update the roles names once it is created
3. it may be a breaking change when updating directus if users are doing point 1 and 2.

### Alternatives

- Writing an script or extension that check for unique in the names, then later check the stage environment roles and permissions, and do the match with the current roles in the production production environment based on the permissions of the stage system

### Adoption Strategy

if you implement the unique of the name, it can be a breaking change as users with repeated roles name will suffer, or users with UUIDs and doing a similar script already as mentioned in the alternative may suffer too if they are expecting the UUID in the permissions response instead of the "name" of the role.

It is possible to write a migration.

If you implement the roles syncs as with the schema, I believe this will just help user to feel safer and deploy faster new instance of directus based on roles and permission pre-tested in a controlled environment.

**Comment by ghost** [comment]
A code-first approach to permissions has been approved, which seems like it would cover your issue: https://github.com/directus/directus/discussions/13041

I'll add though, that in the work my organization has done around roles and keeping them synchronized between environments has been to ensure that we always use the same UUID for each role across environments.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
